### PR TITLE
Add custom location attachment to the Demo App

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Links
Resolves https://github.com/GetStream/ios-issues-tracking/issues/684

### 🎯 Goal
Adds a custom attachment example to the Demo App. This will make it easier to test features for custom attachments. It is also helpful for customers to see how a custom attachment can be implemented. In the future, we should update the documentation to use this example instead of the dummy "workout" example we currently have.

While working on this, I realised it was not easy for custom attachments to support the mixed attachments rendering, so I made some improvements to make it easier. 

### 📝 Summary
- Adds a custom location attachment to the Demo App
- Makes it easier for custom attachments to support mixed attachment rendering
- Adds the possibility to enable/disable background mapping in the demo app.
- Fixes deleted messages, always showing custom attachments

#### Adds a custom location attachment to the Demo App
For this attachment, we are just saving the coordinates of a location. Then, the app is responsible for showing that location on a map. For better performance, an image snapshot of the map is generated in the message list, and only when tapping the snapshot we open the actual map. We could also save the generated snapshots upload it to the CDN, and save them in the message. This way, Web could use the generated snapshots from iOS or Android, but this is out of scope right now, especially for a Demo App example.

#### Makes it easier for custom attachments to support mixed attachment rendering
Previously, in order to support mixed attachments for our custom attachment was quite complex and not intuitive. It was required to override the `MixedAttachmentInjector`, override the `injectors` property, make sure we did it lazily and make sure we defined the correct order. 

Now, we just need to register our custom attachment to the `MixedAttachmentInjector` registry, and then when overriding the Catalog, make sure to return the `MixedAttachmentInjector` when there are multiples attachment types.

**Example:**
```swift
Components.default.mixedAttachmentInjector.register(.location, with: LocationAttachmentViewInjector.self)

class DemoAttachmentViewCatalog: AttachmentViewCatalog {
    override class func attachmentViewInjectorClassFor(message: ChatMessage, components: Components) -> AttachmentViewInjector.Type? {
        let hasMultipleAttachmentTypes = message.attachmentCounts.keys.count > 1
        let hasLocationAttachment = message.attachmentCounts.keys.contains(.location)
        if hasLocationAttachment {
            if hasMultipleAttachmentTypes {
                return MixedAttachmentViewInjector.self
            }
            return LocationAttachmentViewInjector.self
        }

        return super.attachmentViewInjectorClassFor(message: message, components: components)
    }
}
```

#### Fixes deleted messages, always showing custom attachments
We returned `nil` for the attachment injector whenever the message was deleted. But this was done in the incorrect place. Whenever there was a customization, this logic was overriden.

### 🎨 Showcase

https://github.com/GetStream/stream-chat-swift/assets/12814114/3e4bb115-52d1-4fab-b721-be3c2974be35

#### Precondition
Location Attachments should be enabled in the Demo App Configuration. 

#### Adds a custom location attachment to the Demo App
Please check the video above

#### Makes it easier for custom attachments to support mixed attachments rendering
1. Add a location with another attachment
2. Both attachments should render in the composer
3. Both attachments should render in the message list

#### Fixes deleted messages, always showing custom attachments
1. Add a location attachment
2. Delete a message with a location attachment
3. "Message deleted" text should appear.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,9 +10,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Fix editing of async voice messages [#2943](https://github.com/GetStream/stream-chat-swift/pull/2943)
 
 ## StreamChatUI
+### ✅ Added
+- Makes it easier for custom attachments to support mixed attachments rendering [#2947](https://github.com/GetStream/stream-chat-swift/pull/2947)
```
Comment (martinmitrevski) [inline CHANGELOG.md:14]: maybe rephrase this? "makes it" - what's "it"? 
Maybe "Better support for mixed attachments rendering"?

File: Sources/StreamChatUI/ChatMessageList/Attachments/MixedAttachmentViewInjector.swift [hunk 2]
```diff
@@ -3,15 +3,52 @@
...
-        contentView.components.filesAttachmentInjector.init(contentView),
-        contentView.components.voiceRecordingAttachmentInjector.init(contentView)
+    /// The registry of injectors associated with their attachment type, that support mixed attachment rendering.
+    ///
+    /// **Note:** It is an array and not a dictionary, since it defines the order of the rendering of the different types of attachments.
+    /// In order to customise the order, this static property can be changed.
+    ///
+    /// By default, this is the order of how mixed attachments are rendered:
+    ///
+    ///     1. Images and Videos
```
Comment (martinmitrevski) [inline Sources/StreamChatUI/ChatMessageList/Attachments/MixedAttachmentViewInjector.swift:17]: Looking at the code, it's videos then images.

File: docusaurus/docs/iOS/uikit/guides/working-with-attachments.md [hunk 3]
```diff
@@ -270,7 +270,30 @@ Components.default.attachmentViewCatalog = MyAttachmentViewCatalog.self
 
 The `AttachmentViewCatalog` class is used to pick the `AttachmentViewInjector` for a message, in our case we only need to check if the message contains a workout attachment.
 
-If needed you can also send a message with a workout attachment directly from Swift just to test that everything works correctly:
+In case you want your attachments to be rendered a long with other types of attachments, you can do this by adhering your custom injector to the `MixedAttachmentViewInjector`. For this, first, you need to register your custom injector to the `MixedAttachmentViewInjector` registry like so:
```
Comment (martinmitrevski) [inline docusaurus/docs/iOS/uikit/guides/working-with-attachments.md:273]: a long -> along

**Review by martinmitrevski — COMMENTED** [review]
Great stuff @nuno-vieira 👏  Just added few small things, otherwise this is a very nice improvement.

File: DemoApp/StreamChat/Components/CustomAttachments/DemoChatMessageListVC.swift [hunk 4]
```diff
@@ -0,0 +1,7 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import StreamChatUI
+
+class DemoChatMessageListVC: ChatMessageListVC {}
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply DemoApp/StreamChat/Components/CustomAttachments/DemoChatMessageListVC.swift:7]: Right now, we don't have any other customization here, but we do have an implementation in `LocationAttachmentViewDelegate`. The extension is there, so that all location attachment code is inside the Location folder

File: CHANGELOG.md (hunk 1 above)
Comment (nuno-vieira) [reply CHANGELOG.md]: Done ✅

File: docusaurus/docs/iOS/uikit/guides/working-with-attachments.md (hunk 3 above)
Comment (nuno-vieira) [reply docusaurus/docs/iOS/uikit/guides/working-with-attachments.md]: Done ✅

File: Sources/StreamChatUI/ChatMessageList/Attachments/MixedAttachmentViewInjector.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/Attachments/MixedAttachmentViewInjector.swift:17]: This is fine, the order is by the Injector, the type does not really matter here 👍

**Review by martinmitrevski — APPROVED** [review]

File: DemoApp/StreamChat/Components/CustomAttachments/DemoComposerVC.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+//
+
+import StreamChat
+import StreamChatUI
+import UIKit
+
+class DemoComposerVC: ComposerVC {
+    /// For demo purposes the locations are hard-coded.
+    var dummyLocations: [(latitude: Double, longitude: Double)] = [
+        (38.708442, -9.136822), // Lisbon, Portugal
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply DemoApp/StreamChat/Components/CustomAttachments/DemoComposerVC.swift:12]: Done ✅
